presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java

```diff
@@ -53,29 +59,55 @@ public QueryStateInfoResource(
     }
 
     @GET
-    public List<QueryStateInfo> getQueryDiagnosticsByUser(@QueryParam("user") String user)
+    @Produces(MediaType.APPLICATION_JSON)
+    public List<QueryStateInfo> getQueryStateInfos(@QueryParam("user") String user)
     {
-        if (isNullOrEmpty(user)) {
-            return ImmutableList.of();
+        List<QueryInfo> queryInfos = queryManager.getAllQueryInfo();
+
+        if (!isNullOrEmpty(user)) {
+            queryInfos = queryInfos.stream()
+                    .filter(queryInfo -> Pattern.matches(user, queryInfo.getSession().getUser()))
+                    .collect(toImmutableList());
         }
-        List<QueryInfo> userQueries = queryManager.getAllQueryInfo().stream()
-                .filter(queryInfo -> Pattern.matches(user, queryInfo.getSession().getUser()))
+
+        queryInfos = queryInfos.stream()
                 .filter(queryInfo -> !queryInfo.getState().isDone())
                 .collect(toImmutableList());
 
-        Map<ResourceGroupId, ResourceGroupInfo> rootResourceGroupInfos = userQueries.stream()
+        Map<ResourceGroupId, ResourceGroupInfo> rootResourceGroupInfos = queryInfos.stream()
                 .map(queryInfo -> queryManager.getQueryResourceGroup(queryInfo.getQueryId()))
                 .filter(Optional::isPresent)
                 .map(Optional::get)
                 .map(ResourceGroupId::getRoot)
                 .distinct()
                 .collect(toImmutableMap(identity(), resourceGroupManager::getResourceGroupInfo));
 
-        return userQueries.stream()
+        return queryInfos.stream()
                 .map(queryInfo -> createQueryStateInfo(
                         queryInfo,
                         queryManager.getQueryResourceGroup(queryInfo.getQueryId()),
                         queryManager.getQueryResourceGroup(queryInfo.getQueryId()).map(ResourceGroupId::getRoot).map(rootResourceGroupInfos::get)))
                 .collect(toImmutableList());
     }
+
+    @GET
+    @Path("{queryId}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getQueryStateInfo(@PathParam("queryId") String queryId)
```

Comment:
Make the method return the object and throw `WebApplicationException` to handle the error